aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -110,11 +112,19 @@ static TensorIterator make_reduction(
   // efficiency.
   // not generalize this to common mismatched input/output types to avoid cross
   // product of templated kernel launches.
-  if (self.scalar_type() == dtype ||
-      (self.is_cuda() && self.scalar_type() == kHalf && dtype == kFloat)) {
+  if (self.scalar_type() == in_dtype ||
+      (self.is_cuda() && self.scalar_type() == kHalf && in_dtype == kFloat)) {
```

Comment:
Totally agree with what Natalia is saying.

This will basically nuke our math promotion (fp16 input / fp32 output) handling.
CUDA test is probably fine, as any mismatched input/output type would resulted in the code not going through desired path, but with an additional copy kernel (as here: https://github.com/pytorch/pytorch/pull/26181/files#diff-847680695fb0535da86db8b98a13b5a3R119)

Reply:
Since the existing code all has `in_dtype`== `out_dtype`, this change doesn't effect existing code at all. I will change it to `out_dtype` if that makes more sense though.

Although, could the promotion logic just be left left to the machinery added in #22273?

Reply:
I'm confused - there definitely used to be cases where in_dtype != out_dtype (mixed precision reduction) - are you saying that they were effectively disable earlier, in #22273?

Reply:
Taking sum as example here: https://github.com/pytorch/pytorch/blob/fbcb7a9865ede70bd25b739aeb63d3c8ecea5fd8/aten/src/ATen/native/ReduceOps.cpp#L219-L229
The old `dtype` is not necessarily the same as `in_dtype`, considering you do something like `b = fp16_tensor.sum(dtype=torch.float32)`

I don't see the referenced PR #22273 breaking this logic. IIUC, that PR handles type promotion given inputs with mixed data type.

Reply:
> I'm confused - there definitely used to be cases where in_dtype != out_dtype

Previously there was only one variable, `dtype` that was used _both_ as the `dtype` of the output tensor and the `dtype` of the kernel input. Since `argmax` cannot in general use the same `dtype` for both, it has to be split into two seperate variables. If you pass `in_dtype` = `out_dtype` = `dtype` then this is exactly equivalent to the old code and so in that sense there were never cases where `in_dtype != out_dtype`.

> are you saying that they were effectively disable earlier, in #22273?

No, nothing has been disabled. I just meant that if the cast that was on line 117 is meant to be for type promotion, then there is some overlap with the type promotions introduced in #22273 and it might now be redundant.

>I don't see the referenced PR #22273 breaking this logic. IIUC, that PR handles type promotion given inputs with mixed data type.

Unless I'm misunderstanding, `TensorIterator` can be setup to compute the common `dtype` of all operands and perform the operation in that `dtype`. That sounds like what this code was doing before when it cast the input to `dtype`.

Reply:
Ok, I now realized why we are talking past each other. Something changed between 1.2 and current master. In 1.2 there was indeed a mixed precision reduction from fp16 to fp32, so dtype of input was different from dtype of output. On current master, in all cases input is cast to output type. I used the following script
```
import torch

a=torch.randn(1000,1000, dtype=torch.half, device="cuda")
b=a.sum(dim=0, dtype=torch.float)
torch.cuda.synchronize()
```
2 kernels (copy + reduction) are launched on current master, 1 kernel is launched on 1.2.

Reply:
Okay, I think I see where the confusion is. As I had written it, `in_dtype` was not the `dtype` of the input tensor given to `reduce_op` in the half->float special case.

I think 9c709f81fb61822a23609e32224fcb1256853c3b should avoid this confusion.